### DISCUSSION TITLE
Floats cause rounding problems

### DISCUSSION THREAD

**Discussion by JohnG-cc** [CredComSoc/ProtocolWorkingGroup#8] [Reference Implementation]
Financial software universally uses some kind of BigDecimal number type to represent values, since floats are binary numbers and create unavoidable rounding errors when working with decimal fractions:

https://floating-point-gui.de/errors/rounding/

**Comment by JohnG-cc** [comment]
There is a PHP implementation here:

https://github.com/brick/math

**Comment by matslats** [comment]
I didn't think the problem could be so complicated!
I'm still trying to get around the problem in my head.
Let's say there are trunk, branch and twig nodes using seconds, minutes and hours respectively, so the branch unit is worth 60 trunk units and the twig unit is worth 60 branch units (and 3600 trunk units). 
Lets also say the trunk imposes a fee of 1s on a transaction, the smallest divisible unit of the trunk and indeed of the whole tree. This could be represented in the trunk database as 1, but also the branch and twig databases if they are to be accurate, must also store the seconds even if their own internal units are bigger and not normal divisible into seconds
This way, without storing fractions the  twigward nodes need to know the trunkward node's unit in order to know the precision with which to store numbers. And if any exchange rate changes things get very messy.
There's no question in my mind that storage should be denominated in the local unit of account, with 'adequate' precision. Whether the number is a 10 digit float or 10 digit integer feels like a red herring.
1s in minutes is 0.016666666666...
1s in hours is 0.00027777777777...
Unfortunately we can't communicate recurring decimals as a string. I don't know if fractions like 1/3600 can be communicated or even if that would be a good idea. Furthermore we can allow rounding errors between ledgers as long as the string used to generate the hash is the same i.e. the number expressed in the trunkward unit.
So trunk says to branch the fee is 1. because units are always expressed in the trunkward unit (not the trunk unit) and converted by the leafward node. Branch divides by 60 to lets say 5 decimal places 0.01667 and tells that to twig. Twig divides by 60 and stores 0.00027.
thus in this system 5dp seems more than adequate to store seconds in hours, and 4dps enough for the protocol
So does that mean that the protocol should specify the precision?
Does that mean there is an advantage to storing everything as integers i.e. x10000?

**Comment by utunga** [comment]
Short version is that yes there are many advantages to storing everything as integers and then being precise about rounding. I 100% agree that you don't want floats anywhere near financial data. 

@JohnG-cc  linked, in his first comment to a good link about what's wrong with floats. But yeah predictability and precision are chief advantages of using integer math.  Floats just seem to behave in odd ways to humans and/or in 10 digit format, even if they 'make sense' in binary floating point math terms. 

Just as an example, if you pay .3 dollars, say, for two products each worth .1 and .2 dollars. Then you may have some code that does this 

```
0.1 + 0.2 == 0.3  ->  false
```

Which is due to the fact that 
```
0.1 + 0.2         ->  0.30000000000000004
```

But in practice the user might see something saying they haven't paid when they have, or whatever. And then you might think oh well, this can be easily fixed by rounding everything, and yes, it kinda can - fix that particular problem. But then you'll get all sorts of problems, primary amongst them being *when* to do the rounding. Cause as you know you want to round things as *late* as possible, in order to lose the least amount of data but yeah if you are checking if numbers add up correctly somewhere lower down you need to be sure to round them before checking but keep the underlying numbers to pass further up the chain. 

It might seem like a contrived example, and it kind of is. But here's the thing, the real bug that you will get from such things is gonna be *way* more subtle and annoying and hard to understand than this. 

It's just an ugly experience for all involved so i think that's why nobody stores financial data as floats. [1]

So for example I think we should manipulate and process the above numbers as, basically,

```
1000 + 2000 == 3000
```

But then - at display time - display '0.1' and '0.2' and '0.3' respectively

Well, to be a bit more precise **there are a bunch of contexts in which we need to decide what to do** 

#### Contexts we need to understand:
- When **at rest** in our local storage (eg a db)
- When **being manipulated** (this includes rounding rules, for business logic etc)
- When **being transferred** to another server (interchange format, ie json)
- When **being displayed** to the user (including for example csv exports etc)

Then I suggest that 
- For every currency be sure define and name a **currency unit** (for display default purposes) which can be displayed in decimal (not float) format.
- And also define a 'satoshi' for that currency, ie a **higher precision representation** which is stored / processed as an int64.
- Default the higher precision currency unit to be, say, 6 decimal places higher precision than the default currency unit (ie a million times more precise, cause why not)


### My recommendation would be
- When **at rest** we store in higher precision format. We can't (or I dont think we should) actually dictate how implementations store data at rest but I think we can build our reference implementation this way, recommend this, and also publish the required info as part of the required currency definition for any currency.

- When **being manipulated** people SHOULD use the higher precision format, converting to 'treeward' currency first. As far as rounding goes, it's kinda overkill but ideally you would use  bank rounding for any partial amounts - immediately. (Banker's rounding takes .5 up half the time and .5 down the other half of the time, in predictable way).

- When **being displayed** we display everything as a decimal in the relevant currency with appropriate number of d.p.  depending on context. So, for example a user might see `'22.3 honey'` or `'0.0003 dundee volts`' or whatever or be able to type in '0.02 honey' for this thing. (Note that I say decimal instead of float here because we divide out the underlying high precision number right before display. If the language allows decimal then use that).

- When **being transferred** for instance in json messages or REST data - I suggest we also use the display format rules.  JSON schema allows both 'numeric' (aka float) and 'integer' number formats but we can add some more precise rules here about how to represent numbers in json. Ironically (?) I think we would need to use the use numeric aka float type for currency amounts in the schema, but with any luck we can summarize some of the above points to basically say these numbers should not exceed the underlying precision of that currency on the wire, basically so as to avoid sending any numbers like ``` 0.30000000000000004``` over the wire. 

That would be my recommendation anyway... 

[1] In terms of what is common practice, I guess I worked in finance (which is not quite the same as banking) for many years and when designing a system which needed to deal with numbers at many different levels of precision (sometimes as thousands, millions or even billions but also sometimes just small numbers, like earnings per share) and many currencies from Yen to CHF etc  we just decided (after some careful thought) to use 64bit signed integers for anything and everything and convert things at display time. So, for example, if an analyst typed something like 20 into a field that was expecting things in 'millions' they would see '20' and type '20'  but then we would store it as 20,000,000. This worked really well.

I also worked at the reserve bank in NZ for a bit and certainly never saw any floats being used to store any account type data.  Again, they mostly used 64bit integers at rest. Sometimes you would see a 'Decimal' in the code but of course a 'Decimal' (in c#) is really just a glorified integer that handles the dividing out and then multiplying out of a 'display' number' into the underlying integer format before being sent over the wire or to storage.

I will admit that I have never yet came across a situation where it was necessary to use an 'arbitrary precision number' type like Brick/Math though I have always kind of wanted to ;-)

**Comment by matslats** [comment]
Thanks Miles this gets me a long way.
I think what I need to do then is as follows.
Storage & manipulation: Multiply all numbers by 10^7 in the api and in the database; change the reference implementation db fields to int64
Display: Divide by 10^7 and round before displaying.
Transport; Change the API to expect an integer which is 10^7
I'll attempt this next...

**Reply by matslats** [reply]
So far we've probably assumed that all currencies are divisible as decimals, but in my Drupal module I don't assume that - I do all the accounting in the smallest unit of the node, and then format at display time. this allows me to store 150 minutes and the display them "2hrs 30mins" OR to store 150 cents and format them as $1.50
In Drupal I have full control over the transaction form widget so I can build the quant field in the right format. But here the transaction form is external, and the API currently expects the raw number - an integer, the number of cents or minutes.
SO that leave the app builder with 2 choices:
- put a default quant field [ ] cents (or whatever the smallest unit is called)
- construct a field that takes say, a number up to 2 decimal places and convert it to cents before submitting.
Seems reasonable?
So here are the number formats I'm currently working with:
- 'raw' meaning cents or minutes, which is the smallest unit the system can display, however it is not actually an integer because it contains decimal places too small to be rendered in the display format e.g. fractions of a cent or seconds. All amounts in the php and in the api are raw unless stated otherwise.
- 'stored' which is raw * 10^7
- 'display', which is the raw value, formatted according to my special format.

**Reply by utunga** [reply]
You wrote:
> Transport; Change the API to expect an integer which is 10^7

Just for clarity it wasn't my suggestion that we change the API. We can still use default units but in decimal format and with no more precision than say 7 d.p. 

This is more about implementation details than API changes.

**Comment by matslats** [comment]
Yes in my last frenzy the API expects a float instead of an integer, huge numbers are just for storage.

**Comment by dilgreen** [comment]
It seems to me that we can do something with the 'exchange rate', to deal with the issue you raise about hh:mm:ss accounting in one ledger vs decimal in another.
The 'balance of trade' special account could ask the trunkward ledger about its unit of account (it will need to do this in order for admins in the leaf ledger to know what the exchange rate they are setting means).
If there is a required translation, then it can be determined at this point, and possibly even configured in detail (ie decisions about rounding can be made overt, and chosen).

**Comment by matslats** [comment]
there is already an API method for one node to convert to the units of another. This could be used for pricing.
You can ask for the exchange rate with a remote node and each intermediate node does the math with respect to its own unit, and the response is a float.
Exchange rate settings shouldn't be done via the API though. Each node's governors should know the value of the trunkward unit because their own unit is expressed as a ratio of that.

**Reply by dilgreen** [reply]
I wasn't suggesting exchange rate settings through API - they are a local config issue.
I think perhaps I am not sufficiently versed in the detail of this discussion, so I'll revert to lurking mode here.
